### PR TITLE
Add Long File Name Support

### PR DESCRIPTION
**Pull request by NoMoreFood**
- Structures like dirent's 'd_name presume a static size for a path.  To support long names, PATH_MAX, previously defined as MAX_PATH (260), is increased to 32768 everywhere it is used.  Similarly, MAX_PATH was changed to PATH_MAX everywhere it was used.
- Adjusted manifest to opt-in for long file name support and associated manifest file with all executable project files.
- Note: This still requires the following registry key to be set HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled = 1 (does not require reboot)

### CODE REVIEW

**Comment by WSLUser** [comment]
Can we merge this in with as part of the 8.1 release? (git rebase required)

**Comment by ghost** [comment]
quoting from https://github.com/PowerShell/Win32-OpenSSH/issues/1480#issuecomment-541217996
> 
> @manojampalam I put in a pull request ([PowerShell/openssh-portable#400](https://github.com/PowerShell/openssh-portable/pull/400)) to utilize opt-in long file path support and expand the internal buffers to utilize it. This still requires HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled = 1 (does not require reboot). Alternatively we could prepend \?\ to every path in a Win32 API call, but I suspect that would require quite a few more changes.
> 

so mind adding this to the pull request description? as it maybe confusing for even intermediate level users. I mean clarify that users still need to set the environment variable to enable long paths.
----

merged from second comment.

also update the file for testing this change. its faster than users compiling it or more likely, not compiling it.

for the current file, for now, see https://github.com/PowerShell/Win32-OpenSSH/issues/1480#issuecomment-541860559
----
**peace out**
_ghostArch_

**Comment by ghostArch** [comment]
also update the file for testing this change. its faster than users compiling it or more likely, not compiling it.

for the current file from Oct 15, see https://github.com/PowerShell/Win32-OpenSSH/issues/1480#issuecomment-541860559
----
**peace out**
_ghostArch_

**Comment by bagajjal** [comment]
@NoMoreFood - I think this requires more changes and extensive testing. For example, is_absolute_path() requires changes to consider long paths \\?\D:\test_path. With the current logic, we treat this as not absolute path.

**Comment by NoMoreFood** [comment]
@bagajjal The function is_absolute_path() would never practically see a path in the form of \?\D:\Path (nor any other of the many alternate path formats).  The point of enabling long file name support (via the registry directly or by group policy) is so you *don't* have to use that convoluted format.  Sorry if that wasn't clear.

File: contrib/win32/openssh/keygen.vcxproj [hunk 1]
```diff
@@ -264,6 +270,9 @@
       <EntryPointSymbol>wmainCRTStartup</EntryPointSymbol>
       <AdditionalOptions>/debug /debugtype:cv,fixup /opt:ref /opt:icf /incremental:no %(AdditionalOptions)</AdditionalOptions>
     </Link>
+    <Manifest>
+      <AdditionalManifestFiles>targetos.manifest</AdditionalManifestFiles>
```
Comment (bagajjal) [inline contrib/win32/openssh/keygen.vcxproj]: @NoMoreFood - Where did you get this information?

As per [microsoft documentation page](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file),
image

File: contrib/win32/openssh/scp.vcxproj [hunk 2]
```diff
@@ -1,387 +1,405 @@
...
-    </Link>
-  </ItemDefinitionGroup>
-  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
-  <ImportGroup Label="ExtensionTargets">
-  </ImportGroup>
-</Project>
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="paths.targets" />
+  <ItemGroup Label="ProjectConfigurations">
```
Comment (bagajjal) [inline contrib/win32/openssh/scp.vcxproj:4]: This looks like a CRLF issue.. Can you make sure it only shows what has modified?

File: contrib/win32/openssh/sftp.vcxproj [hunk 3]
```diff
@@ -1,391 +1,409 @@
...
-      <EntryPointSymbol>wmainCRTStartup</EntryPointSymbol>
-      <FullProgramDatabaseFile>true</FullProgramDatabaseFile>
-      <AdditionalOptions>/debug /debugtype:cv,fixup /opt:ref /opt:icf /incremental:no %(AdditionalOptions)</AdditionalOptions>
-    </Link>
-  </ItemDefinitionGroup>
-  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
-  <ImportGroup Label="ExtensionTargets">
-  </ImportGroup>
-</Project>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bagajjal) [inline contrib/win32/openssh/sftp.vcxproj:1]: Please do the same thing here. show only the changes required.

File: contrib/win32/openssh/keygen.vcxproj [hunk 4]
```diff
@@ -195,6 +195,9 @@
       <EntryPointSymbol>wmainCRTStartup</EntryPointSymbol>
       <AdditionalOptions>/debug /debugtype:cv,fixup /opt:ref /opt:icf /incremental:no %(AdditionalOptions)</AdditionalOptions>
     </Link>
+    <Manifest>
```
Comment (bagajjal) [inline contrib/win32/openssh/keygen.vcxproj:198]: why is this change no there for ARM64 platform?

Comment (bagajjal) [reply contrib/win32/openssh/keygen.vcxproj:198]: looks like all the proj files here don't have this change for ARM64 platform.

**Comment by bagajjal** [comment]
> @bagajjal The function is_absolute_path() would never practically see a path in the form of ?\D:\Path (nor any other of the many alternate path formats). The point of enabling long file name support (via the registry directly or by group policy) is so you _don't_ have to use that convoluted format. Sorry if that wasn't clear.

SFTP will call is_absolute_path()..
In auth.c, we have windows logic to expand authorized_keys file

File: contrib/win32/openssh/scp.vcxproj (hunk 2 above)
Comment (NoMoreFood) [reply contrib/win32/openssh/scp.vcxproj:4]: I will confirm, but I believe Visual Studio corrected this to make it consistent.

**Comment by NoMoreFood** [comment]
Nothing about is_absolute_path() has to change.  That's the "magic" of LongPathsEnabled = 1 is that all the normal file APIs (e.g., CreateFile, GetFinalPathNameByHandle) suddenly support long file names *without* needing to make any changing to the path syntax.  The only reason is this not the default Windows behavior is for interoperability with other programs (like Windows Explorer) and that developers have legacy code with limited buffer sizes.

> > @bagajjal The function is_absolute_path() would never practically see a path in the form of ?\D:\Path (nor any other of the many alternate path formats). The point of enabling long file name support (via the registry directly or by group policy) is so you _don't_ have to use that convoluted format. Sorry if that wasn't clear.
> 
> SFTP will call is_absolute_path()..
> In auth.c, we have windows logic to expand authorized_keys file

File: contrib/win32/openssh/scp.vcxproj (hunk 2 above)
Comment (NoMoreFood) [reply contrib/win32/openssh/scp.vcxproj:4]: You're right, but the previous file was a *mix* of both CRLF and LF which is definitely not great.  This is probably due people using the GitHub interface to make adjustments.  My file consistent CRLF (Windows Format) like Visual Studio does by default.  You can view the changes in something like Beyond Compare.

File: contrib/win32/openssh/sftp.vcxproj (hunk 3 above)
Comment (NoMoreFood) [reply contrib/win32/openssh/sftp.vcxproj:1]: See above comment.

File: contrib/win32/openssh/keygen.vcxproj (hunk 4 above)
Comment (NoMoreFood) [reply contrib/win32/openssh/keygen.vcxproj:198]: I will look at this.

Comment (bagajjal) [reply contrib/win32/openssh/keygen.vcxproj:198]: I will take care of this.

**Pull request merged**